File: backend/resolver.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/chromedp/chromedp"
)

// Manifest is the bridge between resolver and downloader.
// The resolver writes it; the downloader reads it.
type Manifest struct {
	ID            string         `json:"id"`
	Store         string         `json:"store"`
	CoverImageURL string         `json:"cover_image_url"`
	Pages         []ManifestPage `json:"pages"`
}

// ManifestPage holds the resolved image URL for a single catalog page.
type ManifestPage struct {
	Number   int    `json:"number"`
	ImageURL string `json:"image_url"`
}

// imageExtractionJS finds the main catalog image on the current page.
const imageExtractionJS = `
	(() => {
		// Prefer images inside the current-page container
		const currentPages = document.querySelectorAll('.page--current img, [class*="page--current"] img');
		const candidates = currentPages.length > 0
			? Array.from(currentPages)
			: Array.from(document.querySelectorAll('img')).filter(img =>
				!img.closest('nav') && !img.closest('aside') &&
				!img.closest('.cuprins') && !img.closest('[class*="sidebar"]') &&
				!img.closest('[class*="thumbnail"]'));

		const good = candidates.filter(img => {
			const w = img.naturalWidth || img.width || 0;
			const h = img.naturalHeight || img.height || 0;
			return img.complete && w > 500 && h > 500 &&
				img.src && !img.src.includes('data:image') &&
				!img.src.includes('rs:fit:400') && !img.src.includes('rs:fit:200');
		});

		if (good.length > 0) {
			return JSON.stringify({success: true, url: good[0].src});
		}
		return JSON.stringify({success: false});
	})()
`

// ResolveManifest creates a manifest.json for a given scraper config.
// It launches a headless browser, visits every page URL, and extracts direct image URLs.
func ResolveManifest(configPath string) (*Manifest, error) {
	config, err := LoadScraperConfig(configPath)
	if err != nil {
		return nil, fmt.Errorf("failed to load config: %v", err)
	}

	log.Printf("[resolver] starting for %s", config.ID)

	firstPageNum, err := extractPageNumber(config.FirstPage)
	if err != nil {
		return nil, fmt.Errorf("failed to parse first page number: %v", err)
	}
	lastPageNum, err := extractPageNumber(config.LastPage)
	if err != nil {
		return nil, fmt.Errorf("failed to parse last page number: %v", err)
	}

	// Fixed 800px viewport — reliable single-page rendering on mobile-optimized viewers.
	ctx, cancel := context.WithTimeout(context.Background(), 300*time.Second)
	defer cancel()

	taskCtx, taskCancel := createBrowserContext(ctx, 800, 1200)
	defer taskCancel()

	// Derive store name from config ID (e.g. "lidl-09-02-15-02-2026" → "lidl")
	store := config.ID
	if idx := strings.Index(config.ID, "-"); idx > 0 {
		store = config.ID[:idx]
	}

	manifest := &Manifest{
		ID:    config.ID,
		Store: store,
	}

	// Resolve cover image
	log.Printf("[resolver] resolving cover image: %s", config.CoverImage)
	coverURL, err := extractImageFromPage(taskCtx, config.CoverImage)
	if err != nil {
		log.Printf("[resolver] warning: cover image failed: %v", err)
	} else {
		manifest.CoverImageURL = coverURL
	}

	// Resolve each page
	log.Printf("[resolver] resolving pages %d–%d", firstPageNum, lastPageNum)
	for pageNum := firstPageNum; pageNum <= lastPageNum; pageNum++ {
		pageURL := buildPageURL(config.FirstPage, pageNum)
		log.Printf("[resolver] page %d/%d: %s", pageNum, lastPageNum, pageURL)

		imageURL, err := extractImageFromPage(taskCtx, pageURL)
		if err != nil {
			log.Printf("[resolver] warning: page %d failed: %v", pageNum, err)
			continue
		}

		manifest.Pages = append(manifest.Pages, ManifestPage{
			Number:   pageNum,
			ImageURL: imageURL,
		})

		// Small delay between pages to be respectful
		time.Sleep(500 * time.Millisecond)
	}

	// Write manifest to disk
	baseDir := filepath.Join("../newsletters", config.ID)
	if err := os.MkdirAll(baseDir, 0755); err != nil {
		return nil, fmt.Errorf("failed to create directory: %v", err)
	}

	manifestPath := filepath.Join(baseDir, "manifest.json")
	data, err := json.MarshalIndent(manifest, "", "  ")
	if err != nil {
		return nil, fmt.Errorf("failed to marshal manifest: %v", err)
	}
	if err := os.WriteFile(manifestPath, data, 0644); err != nil {
		return nil, fmt.Errorf("failed to write manifest: %v", err)
	}

	log.Printf("[resolver] manifest written to %s (%d pages resolved)", manifestPath, len(manifest.Pages))
	return manifest, nil
}

// --- helpers ---

// createBrowserContext creates a chromedp context with the specified viewport size.
func createBrowserContext(ctx context.Context, width, height int) (context.Context, context.CancelFunc) {
	opts := append(chromedp.DefaultExecAllocatorOptions[:],
		chromedp.Flag("headless", true),
		chromedp.Flag("disable-gpu", true),
		chromedp.Flag("no-sandbox", true),
		chromedp.Flag("disable-cache", true),
		chromedp.Flag("disable-application-cache", true),
		chromedp.Flag("disk-cache-size", "0"),
		chromedp.WindowSize(width, height),
	)

	allocCtx, allocCancel := chromedp.NewExecAllocator(ctx, opts...)
	taskCtx, taskCancel := chromedp.NewContext(allocCtx)

	cancelFunc := func() {
		taskCancel()
		allocCancel()
	}

	return taskCtx, cancelFunc
}

// extractImageFromPage navigates to a page and extracts the main image URL.
func extractImageFromPage(ctx context.Context, pageURL string) (string, error) {
	var result string

	err := chromedp.Run(ctx,
		chromedp.Navigate(pageURL),
		chromedp.WaitReady("body"),
		chromedp.WaitVisible("img", chromedp.ByQuery),
		chromedp.Sleep(3*time.Second),
		chromedp.Evaluate(imageExtractionJS, &result),
	)
	if err != nil {
		return "", err
	}

	var extractResult struct {
		Success bool   `json:"success"`
		URL     string `json:"url"`
	}
	if err := json.Unmarshal([]byte(result), &extractResult); err != nil || !extractResult.Success {
		return "", fmt.Errorf("no image found on page")
	}

	imageURL := extractResult.URL
	if !strings.HasPrefix(imageURL, "http") {
		parsedURL, err := url.Parse(pageURL)
		if err == nil {
			imageURL = fmt.Sprintf("%s://%s%s", parsedURL.Scheme, parsedURL.Host, imageURL)
		}
	}

	return imageURL, nil
}

// extractPageNumber extracts the page number from a URL like ".../page/42".
func extractPageNumber(pageURL string) (int, error) {
	re := regexp.MustCompile(`/page/(\d+)`)
	matches := re.FindStringSubmatch(pageURL)
	if len(matches) < 2 {
		return 0, fmt.Errorf("page number not found in URL: %s", pageURL)
	}
	return strconv.Atoi(matches[1])
}

// buildPageURL replaces the page number in a template URL.
func buildPageURL(templateURL string, pageNum int) string {
	re := regexp.MustCompile(`/page/\d+`)
	return re.ReplaceAllString(templateURL, fmt.Sprintf("/page/%d", pageNum))
}

File: src/resolver.ts
import { chromium, type Page } from "playwright";
import { readdir, readFile, mkdir, writeFile } from "node:fs/promises";
import { join } from "node:path";

// --- Types ---

interface ScraperConfig {
  id: string;
  cover_image: string;
  first_page: string;
  last_page: string;
}

export interface ManifestPage {
  number: number;
  image_url: string;
}

export interface Manifest {
  id: string;
  store: string;
  cover_image_url: string;
  pages: ManifestPage[];
}

// --- Image extraction (runs inside the browser) ---

function extractMainImage(): { success: boolean; url?: string } {
  const currentPages = document.querySelectorAll(
    ".page--current img, [class*='page--current'] img"
  );
  const candidates =
    currentPages.length > 0
      ? Array.from(currentPages) as HTMLImageElement[]
      : (Array.from(document.querySelectorAll("img")) as HTMLImageElement[]).filter(
          (img) =>
            !img.closest("nav") &&
            !img.closest("aside") &&
            !img.closest(".cuprins") &&
            !img.closest('[class*="sidebar"]') &&
            !img.closest('[class*="thumbnail"]')
        );

  const good = candidates.filter((img) => {
    const w = img.naturalWidth || img.width || 0;
    const h = img.naturalHeight || img.height || 0;
    return (
      img.complete &&
      w > 500 &&
      h > 500 &&
      img.src &&
      !img.src.includes("data:image") &&
      !img.src.includes("rs:fit:400") &&
      !img.src.includes("rs:fit:200")
    );
  });

  if (good.length > 0) {
    return { success: true, url: good[0]!.src };
  }
  return { success: false };
}

// --- Helpers ---

function extractPageNumber(url: string): number {
  const match = url.match(/\/page\/(\d+)/);
  if (!match) throw new Error(`Page number not found in URL: ${url}`);
  return parseInt(match[1]!, 10);
}

function buildPageURL(templateURL: string, pageNum: number): string {
  return templateURL.replace(/\/page\/\d+/, `/page/${pageNum}`);
}

async function extractImageFromPage(
  page: Page,
  url: string
): Promise<string> {
  await page.goto(url, { waitUntil: "domcontentloaded" });
  await page.waitForSelector("img", { timeout: 15000 });
  // Give images time to load their full src
  await page.waitForTimeout(3000);

  const result = await page.evaluate(extractMainImage);

  if (!result.success || !result.url) {
    throw new Error("No image found on page");
  }

  let imageURL = result.url;
  if (!imageURL.startsWith("http")) {
    const parsed = new URL(url);
    imageURL = `${parsed.protocol}//${parsed.host}${imageURL}`;
  }

  return imageURL;
}

// --- Main resolver ---

export async function resolveManifest(configPath: string): Promise<Manifest> {
  const raw = await readFile(configPath, "utf-8");
  const config: ScraperConfig = JSON.parse(raw);

  console.log(`[resolver] starting for ${config.id}`);

  const firstPageNum = extractPageNumber(config.first_page);
  const lastPageNum = extractPageNumber(config.last_page);

  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext({
    viewport: { width: 800, height: 1200 },
  });
  const page = await context.newPage();

  try {
    // Derive store name from config ID (e.g. "lidl-09-02-15-02-2026" -> "lidl")
    const store = config.id.includes("-")
      ? config.id.slice(0, config.id.indexOf("-"))
      : config.id;

    const manifest: Manifest = {
      id: config.id,
      store,
      cover_image_url: "",
      pages: [],
    };

    // Resolve cover image
    console.log(`[resolver] resolving cover image: ${config.cover_image}`);
    try {
      manifest.cover_image_url = await extractImageFromPage(
        page,
        config.cover_image
      );
    } catch (err) {
      console.warn(`[resolver] warning: cover image failed: ${err}`);
    }

    // Resolve each page
    console.log(`[resolver] resolving pages ${firstPageNum}-${lastPageNum}`);
    for (let pageNum = firstPageNum; pageNum <= lastPageNum; pageNum++) {
      const pageURL = buildPageURL(config.first_page, pageNum);
      console.log(`[resolver] page ${pageNum}/${lastPageNum}: ${pageURL}`);

      try {
        const imageURL = await extractImageFromPage(page, pageURL);
        manifest.pages.push({ number: pageNum, image_url: imageURL });
      } catch (err) {
        console.warn(`[resolver] warning: page ${pageNum} failed: ${err}`);
      }

      // Small delay between pages to be respectful
      await new Promise((r) => setTimeout(r, 500));
    }

    // Write manifest to disk
    const baseDir = join("newsletters", config.id);
    await mkdir(baseDir, { recursive: true });

    const manifestPath = join(baseDir, "manifest.json");
    await writeFile(manifestPath, JSON.stringify(manifest, null, 2));

    console.log(
      `[resolver] manifest written to ${manifestPath} (${manifest.pages.length} pages resolved)`
    );
    return manifest;
  } finally {
    await browser.close();
  }
}

// --- CLI entry point ---

if (import.meta.main) {
  const configPath = process.argv[2];
  if (!configPath) {
    console.error("Usage: bun run src/resolver.ts <config-path>");
    console.error("Example: bun run src/resolver.ts configs/lidl-09-02-15-02-2026.json");
    process.exit(1);
  }
  resolveManifest(configPath).catch((err) => {
    console.error("[resolver] fatal:", err);
    process.exit(1);
  });
}
